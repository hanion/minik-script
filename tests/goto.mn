// goto.mn

// basic for loop with goto
{
	i := 0;
	label loop;
	if i < 5 {
		print("i =", i);
		++i;
		goto loop;
	}
}

// basic goto usage within a loop
{
	i := 0;
	for ; i < 10; ++i {
		if i == 5 {
			goto end;
		}
	}

	label end;
	print("finished loop at:", i);
}

// multiple goto statements and labels
{
	i := 0;
	label loop_start;
	for ; i < 10; ++i {
		if i == 3 {
			print("jumping to mid");
			goto mid;
		}
		if i == 7 {
			print("jumping to end");
			goto end;
		}
	}

	label mid;
	print("middle i =", i);
	if i == 3 {
		++i;
		goto loop_start;
	}

	label end;
	print("end i =", i);
}

// using goto to break out of nested loops
{
	outer := 0;
	inner := 0;

	label outer_loop_start;
	for ; outer < 5; ++outer {

		label inner_loop_start;
		for ; inner < 5; ++inner {
			if inner == 3 {
				print("breaking out of inner loop");
				goto outer_loop_end;
			}
			print("inner loop:", inner);
		}

		label outer_loop_end;
		print("outer loop:", outer);
	}
}


// goto skipping over code
{
	x := 0;
	label skip;

	if x == 0 {
		goto end;
		print("this should be skipped.");
	}

	label end;
}

// goto with an infinite loop and proper exit
{
	for i := 0; true; ++i {
		if i == 5 {
			print("exiting infinite loop at i = ", i);
			goto end;
		}
		print("i = ", i);
	}
	label end;
}



