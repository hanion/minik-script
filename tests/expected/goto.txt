i = 0.000000
i = 1.000000
i = 2.000000
i = 3.000000
i = 4.000000
finished loop at: 5.000000
jumping to mid
middle i = 3.000000
jumping to end
end i = 7.000000
inner loop: 0.000000
inner loop: 1.000000
inner loop: 2.000000
breaking out of inner loop
outer loop: 0.000000
breaking out of inner loop
outer loop: 1.000000
breaking out of inner loop
outer loop: 2.000000
breaking out of inner loop
outer loop: 3.000000
breaking out of inner loop
outer loop: 4.000000
i =  0.000000
i =  1.000000
i =  2.000000
i =  3.000000
i =  4.000000
exiting infinite loop at i =  5.000000
